sklearn/datasets/_arff_parser.py

```diff
@@ -363,18 +368,37 @@ def _pandas_arff_parser(
             dtypes[name] = "Int64"
         elif column_dtype.lower() == "nominal":
             dtypes[name] = "category"
+    # since we will not pass `names` when reading the ARFF file, we need to translate
+    # `dtypes` from column names to column indices to pass to `pandas.read_csv`
+    dtypes_positional = {
+        col_idx: dtypes[name]
+        for col_idx, name in enumerate(openml_columns_info)
+        if name in dtypes
+    }
 
-    # ARFF represents missing values with "?"
-    frame = pd.read_csv(
-        gzip_file,
-        header=None,
-        na_values=["?"],  # missing values are represented by `?`
-        comment="%",  # skip line starting by `%` since they are comments
-        quotechar='"',  # delimiter to use for quoted strings
-        names=[name for name in openml_columns_info],
-        dtype=dtypes,
-        skipinitialspace=True,  # skip spaces after delimiter to follow ARFF specs
-    )
+    default_read_csv_kwargs = {
+        "header": None,
+        "index_col": False,  # always force pandas to not use the first column as index
+        "na_values": ["?"],  # missing values are represented by `?`
+        "comment": "%",  # skip line starting by `%` since they are comments
+        "quotechar": '"',  # delimiter to use for quoted strings
+        "skipinitialspace": True,  # skip spaces after delimiter to follow ARFF specs
+        "escapechar": "\\",
+        "dtype": dtypes_positional,
+    }
+    read_csv_kwargs = {**default_read_csv_kwargs, **(read_csv_kwargs or {})}
+    frame = pd.read_csv(gzip_file, **read_csv_kwargs)
```

Comment:
> we cannot make the retry within the _arff_parser.py file because we get potentially a HTTPResponse where we cannot move to the start of the file. 

At this point the gzip file is on the file system, and there is no more `HTTPResponse` response. In that sense, we can `seek` back to the original file. Are you thinking that the data itself can be corrupted, so retrying here would be a waste?

Reply:
I don't think that the file can be corrupted because we would already have caught it with the checksum indeed.

From what I recall from 2 weeks ago is that you can get a `_MockHTTPResponse` in the test as a `gzip_file` and this object does not expose any `seek` method. But actually looking at `HTTPResponse`, it should implement the `seek` method if the file allows for random access.

I am still unsure if going to the inference/sniffing road would be wise or if we just rely on `ParseError` from Pandas. @thomasjpfan WDYT?

Reply:
After looking over this PR again, I am okay with how it works in this PR.